components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/websocket/InboundWebsocketSourceHandler.java

```diff
@@ -231,19 +264,60 @@ private void handleWebSocketFrame(ChannelHandlerContext ctx, WebSocketFrame fram
+                        contentType = defaultContentType;
+                    }
+
                     handleWebsocketBinaryFrame(frame);
+
+                    org.apache.axis2.context.MessageContext axis2MsgCtx =
+                            ((org.apache.synapse.core.axis2.Axis2MessageContext) synCtx).getAxis2MessageContext();
+
+                    Builder builder = BuilderUtil.getBuilderFromSelector(contentType, axis2MsgCtx);
+                    if (builder != null && InboundWebsocketConstants.BINARY_BUILDER_IMPLEMENTATION.equals(
+                            builder.getClass().getName())) {
+                        synCtx.setProperty(InboundWebsocketConstants.WEBSOCKET_BINARY_FRAME_PRESENT, true);
+                    } else {
+                        synCtx.setProperty(InboundWebsocketConstants.WEBSOCKET_BINARY_FRAME_PRESENT, false);
+                    }
+                    InputStream in = new AutoCloseInputStream(new ByteBufInputStream(frame.content()));
+                    OMElement documentElement = builder.processDocument(in, contentType, axis2MsgCtx);
+                    synCtx.setEnvelope(TransportUtils.createSOAPEnvelope(documentElement));
+                    injectToSequence(synCtx, endpoint);
                     return;
                 } else if ((frame instanceof TextWebSocketFrame) && ((handshaker.selectedSubprotocol() == null) ||
                         (handshaker.selectedSubprotocol() != null
-                                && !handshaker.selectedSubprotocol().contains(InboundWebsocketConstants.SYNAPSE_SUBPROTOCOL_PREFIX)))) {
+                                && !handshaker.selectedSubprotocol().contains(
+                                        InboundWebsocketConstants.SYNAPSE_SUBPROTOCOL_PREFIX)))) {
+                    String contentType = handshaker.selectedSubprotocol();
+                    if(contentType == null && defaultContentType != null) {
+                        contentType = defaultContentType;
+                    }
                     handleWebsocketPassthroughTextFrame(frame);
+                    org.apache.axis2.context.MessageContext axis2MsgCtx =
+                            ((org.apache.synapse.core.axis2.Axis2MessageContext) synCtx).getAxis2MessageContext();
+                    Builder builder = BuilderUtil.getBuilderFromSelector(contentType, axis2MsgCtx);
+
+                    if (builder != null && InboundWebsocketConstants.TEXT_BUILDER_IMPLEMENTATION.equals(
+                            builder.getClass().getName())) {
+                        synCtx.setProperty(InboundWebsocketConstants.WEBSOCKET_TEXT_FRAME_PRESENT, true);
+                    } else {
+                        synCtx.setProperty(InboundWebsocketConstants.WEBSOCKET_TEXT_FRAME_PRESENT, false);
+                    }
+                    InputStream in = new AutoCloseInputStream(new ByteArrayInputStream(
+                            ((TextWebSocketFrame) frame).text().getBytes()));
```

Comment:
(TextWebSocketFrame) frame).text() should not be used , instead the frame should be duplicated.